### ISSUE TITLE
Payment request buttons Incompatible with WooCommerce Product Bundles

### ISSUE THREAD

**Issue by WillBrubaker** [Automattic/woocommerce-payments#3147]
### Describe the bug

The payment request buttons on a single product page fail to use correct pricing or to add any shipping on a bundled product.

### To Reproduce

1. have a bundled product that would require shipping
2. Try to purchase that via payment request button.
3. See no shipping.
4. Further, if the bundled product uses individual pricing, the price is not calculated correctly

### Actual behavior

No shipping
Price not calculated correctly if using individual pricing.

### Screenshots
![see screen shot](https://d.pr/i/9p5ORy+)
https://d.pr/i/9p5ORy
### Expected behavior
For the price and shipping to be calculated the same as if they are added to the cart

### Desktop (please complete the following information):

* OS: [e.g. iOS]
* Browser [e.g. chrome, safari]
* Version [e.g. 22]

### Smartphone (please complete the following information):

* Device: [e.g. iPhone6]
* OS: [e.g. iOS8.1]
* Browser [e.g. stock browser, safari]
* Version [e.g. 22]

### Additional context

4374404-zen

**Comment by malithsen** [comment]
Some additional context, this problem exists in Payment Request Button (PRB) in Stripe gateway plugin as well. Various solutions have been explored in the context of Strip gatway plugin but as of yet, we don't have a definitive solution. As far as I can see same issues/approaches extends to WCPay as well.

1. PRB configurations are set in backend with a static set of configurations. Once the PRB is initialized, only certain configurations can be updated. `request_shipping` is not one of those that can be dynamically updated.
2. The PRB options are obtained directly from the product and not from the cart.
3. Bundle may contain optional or virtual products, so based on what  customer selects within the product page, bundle might end up not requiring shipping. Because of this, we'll have to tear-down and re-initialize the PRB button  
4. Bundles calculate the final price and shipping requirement after it is added to the cart

To give accurate price/shipping requirement we will have to listen to product configuration changes via event listeners and replicate some(most?) of the processing that Bundles plugin does in WCPay.

Another approach @asumaran has suggested in pca54o-29Y-p2 to introduce temporary carts that are independent of the default cart, so that product information can be obtained without affecting the main cart. 

More discussion on this issue from Stripe Gateway side is available here pcrvl0-e2-p2

**Comment by pierorocca** [comment]
@c-shultz is this a WooPayments issue rather than WooPay which has achieved compatibility?

**Comment by c-shultz** [comment]
>is this a WooPayments issue 

Yeah, this looks like a general WooPayments issue. 

>More discussion on this issue from Stripe Gateway side is available here pcrvl0-e2-p2

@diegocurbelo is this still an active issue on the Stripe plugin? It looks like we might want to prioritize a cross-team/product effort to get this fixed in the same way if so.

**Comment by asumaran** [comment]
@c-shultz I believe this is an issue in the Payment Request API itself. The core of the problem is that we need to get the product info from the Cart every time the user clicks the G Pay/Apple Pay button.

I've tried to implement this approach when working on https://github.com/woocommerce/woocommerce-gateway-stripe/issues/1563 but ended up not working because we couldn't make an asynchronous call after the user clicks the G Pay/Apple Pay button. [^1]

--

Recently, during my parental leave I was thinking about this issue and found out the W3C did an important change to the spec of the Payment Request API. [Now you can pass a promise](https://www.w3.org/TR/payment-request/#show-method) to the `show()` method (used to display the payment sheet). That way we can make an async call and update the payment details before displaying it. This sole change allow us to fix https://github.com/woocommerce/woocommerce-gateway-stripe/issues/1563 and other issues like this one. 

The problem is that the JS Stripe library **doesn't accept** passing a promise to their `show()` [method](https://stripe.com/docs/js/payment_request/show).

I did a quick POC using the Payment Request API and Google Chrome (Google Pay) where we can see it's possible to update the payment sheet after an async call.

https://github.com/Automattic/woocommerce-payments/assets/1025173/666bd582-3969-4b36-8be0-1e6ac951c0a5

- Notice the payment sheet start with `$1` – The parameter `getTotalsFromCart()` passed to the `show()` method is returned early just to avoid returning the promise.
- I've used `setTimeout` to simulate an XHR request.
- When `getTotalsFromCart()` returns the promise the payment sheets waits for the async call to finish and the amount is updated with the `$200` that came from the resolved promise.
- I didn't test with Safari/Apple Pay but I'm optimistic it will also work.

So, while Stripe doesn't update their JS library, fixing issues like this one, where we need data from the backend, will be challenging. As mentioned above by @malithsen, I've suggested using "temporary" carts as a workaround. Besides not being a perfect solution, it'll also require a lot of effort from multiple teams. 

[^1]: Actually, we can make async requests, the issue was Safari throwing an error after 1000ms execution which was unacceptable because we can't guarantee the request is going to take less than that. More info about this: pcrvl0-e2-p2#comment-251

**Comment by pierorocca** [comment]
> So, while Stripe doesn't update their JS library,

Sooner or later @c-shultz we'll need to migrate away from Stripe's PRB element, a W3C implementation that Stripe no longer updates, to the PE element type or possibly the Express Checkout element. Adding new express checkouts beyond GPay and Apple Pay would require it.

**Comment by c-shultz** [comment]
Thanks for the digging and details on this, @asumaran.

>Sooner or later @c-shultz we'll need to migrate away from Stripe's PRB element, a W3C implementation that Stripe no longer updates, to the PE element type or possibly the Express Checkout element.

That will be important to consider so we're solving this in the right place (wouldn't want to fix this for the PRB in a way that's not future proof).

>Actually, we can make async requests, the issue was Safari throwing an error after 1000ms execution which was unacceptable because we can't guarantee the request is going to take less than that.

@asumaran I see the click event for the ExpressCheckoutElement has as on click callback but it also mentions a 1000ms time limit: https://stripe.com/docs/js/element/events/on_click?type=expressCheckoutElement

I don't see any `show()` events for the expressCheckoutElement, but I may not be looking in the right place. Do you think we'll run into the same limitation with this implementation?

>I've suggested using "temporary" carts as a workaround. Besides not being a perfect solution, it'll also require a lot of effort from multiple teams.

As I was catching up on the discussion there, I saw that the cart tokens may go a long way towards this as you discussed here with @dechov: pca54o-29Y-p2#comment-4564 Do you believe there's still promise in this direction and this likely something we would just need to prioritize? If I'm understanding correctly, the cart tokens could allow temporary cart to be created with a new token utilizing existing WooCommerce core functionality?

**Comment by c-shultz** [comment]
It looks like this issue has the same root cause, @asumaran? https://github.com/Automattic/woocommerce-payments/issues/3417 Are you aware of any other issues on WooPayments?

And also most of the open issues listed for Stripe here could be solved with temporary carts? https://github.com/woocommerce/woocommerce-gateway-stripe/issues/1563 Or am I overgeneralizing on this?

**Comment by c-shultz** [comment]
There's a tangled web of overlapping concerns among different PRB issues. Even the custom field discussion for PRBs (https://github.com/Automattic/woocommerce-payments/issues/6539#issuecomment-1792452974) touches on the management of shopping carts:
>In the past, I found [this payment button implementation](https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/blob/03c9b5d24db5983bca7d461c17a1674ab7e70075/assets/js/wc-gateway-ppec-generate-cart.js) to be the most robust one and the best candidate for inclusion in a "Core API" (product context). The code was equally well-mannered in the cart/checkout. Check out how it relays form field validation to Woo Core -- behind the scenes, and if I remember correctly, it "simulates" an add-to-cart operation and validates cart contents. Imagine having a shared foundation like this in Core for all express checkout buttons to use.

**Comment by pierorocca** [comment]
Clarification, Payment Elements are for external displaying and redirecting to 3P payments options. They are not integrated fully with Stripe. Express Checkout Element is the new element we'd want to migrate to, to unlock additional payment methods and the latest and greatest features. https://stripe.com/docs/elements/express-checkout-element/migration

Thanks @c-shultz for looking at future state to see if it would have the same issue. 👍

**Comment by asumaran** [comment]
> That will be important to consider so we're solving this in the right place.

I didn't think of making our own library to handle express checkouts before. I don't know if we have the obligation to use the JS Stripe library, if not, that seems like a good idea given that we depend on the Stripe JS library to fix many issues related to Payment Request buttons. We need to consider the cost and benefits on doing so since the JS Stripe library seems to be a wrapper of the [Apple Pay JS API ](https://developer.apple.com/documentation/apple_pay_on_the_web/apple_pay_js_api) and the Payment Request API.

>but it also mentions a 1000ms time limit:

I don't remember seeing that warning in their documentation. They may have stumbled upon into the same scenario as us.

>I don't see any show() events for the expressCheckoutElement

The `.show()` method it's only used when the event propagation is stopped otherwise the payment sheet will be displayed automatically. In [WCStripe](https://github.com/woocommerce/woocommerce-gateway-stripe) we stop the event propagation [here](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/assets/js/stripe-payment-request.js#L638) and call the `.show()` method [here](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/7a8b2bcaad99dd2bf573dc6b74430c761120ff95/assets/js/stripe-payment-request.js#L664) and [here](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/7a8b2bcaad99dd2bf573dc6b74430c761120ff95/assets/js/stripe-payment-request.js#L784) depending of the context. I assume it's very similar on WooPayments.

>I saw that the cart tokens may go a long way towards this

Yes it does. The approach with it will be very different though. We may also need to tweak the store API since creating multiple carts from the browser is _not possible_ IIRC since the browser sends the cookies automatically on all XHR requests ([mentioned](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2461) in the caveats of the POC ), so WC will always use the default cart.

>It looks like this issue has the same root cause? https://github.com/Automattic/woocommerce-payments/issues/3417 Are you aware of any other issues on WooPayments?

Yes it's the same issue. I believe WC Stripe and WooPayments share the same PRB issues.

**Comment by c-shultz** [comment]
I kicked off a thread in Slack (p1705607994165039-slack-C043JN7AJHY). I think the only path forward here is to address this as part of a bigger picture project. We've seen the individual fixes get repeatedly blocked because it's hard to justify the investment to fix a single issue or two.

**Comment by gpressutto5** [comment]
I've been trying to reproduce this issue with Bundles that require shipping and individual pricing, but everything seems to be working fine. (Ignore the problem with the cents. It's a rounding problem in my environment and does not require a bundle or PRB to happen)

| | Cart | Product Page | PRB |
| ---- | ---- | ---- | ---- |
| <b>Apple Pay</b> | <img width="814" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/12385501/ca1d31f3-0edc-469b-aa38-e46f681a47a0"> | <img width="470" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/12385501/5829485f-f7f6-4643-8525-04bfd6334196"> | <img width="405" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/12385501/12746e97-4f2a-4627-bbec-410ffa6434b4"> |
| <b>Google Pay</b> | <img width="809" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/12385501/fd10e64a-2f8a-4a69-aff0-bb34dd4add09"> | <img width="721" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/12385501/9a74e360-c317-4f50-985e-437690f0a90d"> | <img width="617" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/12385501/dcea973f-d38c-4dac-b7c1-c656faf9d2b6"> |

Am I missing something?

**Comment by gpressutto5** [comment]
After talking to @asumaran, there are many fragile workarounds in PRB, and a major refactor that uses real cart data would fix most conflicts.
I think we can leave this issue open but blocked until this refactor is done.

**Comment by pierorocca** [comment]
@c-shultz fyi related to planning consideration/SPIKE for possibly migrating the PRB element and getting ahead of Stripe's compliance date.

**Comment by FangedParakeet** [comment]
I have begun a discussion deliberating on the destiny of the PRB implementation in this post -> paJDYF-ckK-p2 FYI

**Comment by helgatheviking** [comment]
Following because this is relevant to Mix and Match Products and Name Your Price as well. Anything that has custom prices or adds multiple products simultaneously has always had an issue with Payment Request buttons and we've had to disable them in lieu of displaying them when they don't work as expected.

**Comment by frosso** [comment]
Now that #10016 has been merged, I believe this has been fixed.

Here's a cart example of a product bundle:

<img width="834" alt="Image" src="https://github.com/user-attachments/assets/477659a6-2648-41df-b6a7-6d17719a15d0" />

Before, this was the summary:

<img width="604" alt="Image" src="https://github.com/user-attachments/assets/063af72e-ed8f-4cbe-8cc9-5889c2247fa6" />

Now, this is the summary:

<img width="587" alt="Image" src="https://github.com/user-attachments/assets/f6fa020b-8dc8-40be-a5eb-adab85bf6139" />

The individual line items of the bundle can't really be displayed in the summary, since they could influence the total amount when summed together (and the GooglePay/ApplePay dialog might complain about the discrepancy).
But at least the amounts (product amount/shipping) are accurate - which was the main point of the ticket.

I'll point out that it's still not possible to place a bundled order w/ ApplePay/GooglePay directly from the product page.

Thank you for the patience with this, closing for now.

**Comment by helgatheviking** [comment]
>I'll point out that it's still not possible to place a bundled order w/ ApplePay/GooglePay directly from the product page.

That's the original issue though... So unless there's a new issue for it, I'm not sure this should be closed.

**Comment by pierorocca** [comment]
@frosso @FangedParakeet is this something that can be addressed in core? I understand the challenge here is that the product, in essence, doesn't exist until the shopper makes their choices of what's contained in the bundle. What would you need from core in order to make this work?

**Comment by frosso** [comment]
> > I'll point out that it's still not possible to place a bundled order w/ ApplePay/GooglePay directly from the product page.

> That's the original issue though... So unless there's a new issue for it, I'm not sure this should be closed.

My apologies, I interpreted the original issue as merely ensuring that the pricing/shipping for the product bundle was displayed correctly in the GooglePay/ApplePay dialog.
I closed this issue after verifying that [the tokenized ECE changes](https://github.com/Automattic/woocommerce-payments/pull/10016) were displaying the correct pricing for shipping/tax on cart & checkout (while the previous PRB/ECE implementation didn't do that).

While digging more deeply at the history of this ticket & reports, I think that at the time of writing this ticket, the GooglePay/ApplePay buttons were displayed on all PDPs for product bundles.
But then, I believe some modifications were made to exclude ApplePay/GooglePay on _some_ PDPs for product bundles.

More specifically, the WC Product Bundles extension itself hooks into the `wcpay_payment_request_is_product_supported` filter (which is triggered on product pages) to decide whether the bundle should display the GooglePay/ApplePay button.

Despite these modifications, the GooglePay/ApplePay buttons were still showing incorrect shipping/tax information.

So, I'd like to rectify my initial comment saying that the GooglePay/ApplePay is displayed on _some_ WC Product Bundles product pages - as demonstrated by this screenshot:

<img width="819" alt="Image" src="https://github.com/user-attachments/assets/9a2df37a-9c64-4b5d-8aa6-74abf786e83b" />

However:
- Some additional fixes are/were needed to make even this scenario _more_ compatible, without issues (I took the liberty of investigating some of those issues and resolving them here: https://github.com/Automattic/woocommerce-payments/pull/10295
- More complex configuration of product bundles are not deemed compatible by the WC Product Bundles plugin itself (via the `wcpay_payment_request_is_product_supported` filter)

<img width="663" alt="Image" src="https://github.com/user-attachments/assets/47f357de-8f07-486d-9b6d-fea15e0855f5" />

> is this something that can be addressed in core? [...] What would you need from core in order to make this work?

WC Product Bundle added Store API support in v8.3.0, just a few weeks ago at the time of this writing.
This feature enables us to add ApplePay/GooglePay support for product bundles on all product pages.
Cart/Checkout support for product bundles via the tokenized ECE will be released on WooPayments 8.9.0.

In terms of WC Core, it would be great to have a way to avoid relying on parsing the DOM to determine which attributes are selected so we can reduce the brittleness of the codebase - but it's a large ask requiring a lot of effort across all extensions.

So far, we've just been parsing the DOM on all the other extensions we support - we can follow the same approach on this use case as well.

To recap:
- Since I fixed some additional compatibility issues w/ some "basic" product bundles configuration [in here (which will likely be released w/ WooPayments 9.0.0)](https://github.com/Automattic/woocommerce-payments/pull/10295) & since the cart/checkout are now fully compatible with WC Product Bundles [due to the tokenized ECE in WooPayments 8.9.0](https://github.com/Automattic/woocommerce-payments/pull/10016), I think we can keep this ticket closed
- I created a separate ticket to add support to more "advanced" WC Product Bundles configuration in https://github.com/Automattic/woocommerce-payments/issues/10296 - since it requires changes on both the WC Product Bundles plugin & WooPayments codebases

What do we think?

**Comment by pierorocca** [comment]
@frosso thanks for the comprehensive recap and for the tokenized based enhancements you've been spearheading since late 2023.

> In terms of WC Core, it would be great to have a way to avoid relying on parsing the DOM to determine which attributes are selected so we can reduce the brittleness of the codebase - but it's a large ask requiring a lot of effort across all extensions.

@ralucaStan could you weigh in on this point /  #10296?

**Comment by helgatheviking** [comment]
>So, I'd like to rectify my initial comment saying that the GooglePay/ApplePay is displayed on some WC Product Bundles product pages - as demonstrated by this screenshot:

Presumably this is shown on bundles that do not require any configuration. 

Can someone tell me what ECE stands for?

**Comment by asumaran** [comment]
> Can someone tell me what ECE stands for?

@helgatheviking Express Checkout Element https://docs.stripe.com/elements/express-checkout-element?locale=en-GB
